File: clsMasicEventNotifier.vb
Imports MASIC.clsMASIC
Imports PRISM

Public MustInherit Class clsMasicEventNotifier
    Inherits EventNotifier

    Private mLastPercentComplete As Short = 0

#Region "Events"

    ''' <summary>
    ''' Provides information on the number of cache and uncache events in objSpectraCache
    ''' </summary>
    ''' <param name="cacheEventCount"></param>
    ''' <param name="unCacheEventCount"></param>
    Public Event UpdateCacheStatsEvent(cacheEventCount As Integer, unCacheEventCount As Integer)

    ''' <summary>
    ''' Update the code associated with an error
    ''' </summary>
    ''' <param name="eNewErrorCode"></param>
    Public Event UpdateBaseClassErrorCodeEvent(eNewErrorCode As ProcessFilesErrorCodes)

    ''' <summary>
    ''' Update the code associated with an error
    ''' </summary>
    ''' <param name="eNewErrorCode"></param>
    ''' <param name="leaveExistingErrorCodeUnchanged"></param>
    Public Event UpdateErrorCodeEvent(eNewErrorCode As eMasicErrorCodes, leaveExistingErrorCodeUnchanged As Boolean)

#End Region

    Private Sub OnUpdateCacheStats(cacheEventCount As Integer, unCacheEventCount As Integer)
        RaiseEvent UpdateCacheStatsEvent(cacheEventCount, unCacheEventCount)
    End Sub

    Private Sub OnUpdateBaseClassErrorCode(eNewErrorCode As ProcessFilesErrorCodes)
        RaiseEvent UpdateBaseClassErrorCodeEvent(eNewErrorCode)
    End Sub

    Private Sub OnUpdateErrorCode(eNewErrorCode As eMasicErrorCodes, leaveExistingErrorCodeUnchanged As Boolean)
        RaiseEvent UpdateErrorCodeEvent(eNewErrorCode, leaveExistingErrorCodeUnchanged)
    End Sub

    Protected Overloads Sub RegisterEvents(oClass As clsMasicEventNotifier)
        MyBase.RegisterEvents(oClass)

        AddHandler oClass.UpdateCacheStatsEvent, AddressOf UpdatedCacheStatsEventHandler
        AddHandler oClass.UpdateBaseClassErrorCodeEvent, AddressOf UpdateBaseClassErrorCodeEventHandler
        AddHandler oClass.UpdateErrorCodeEvent, AddressOf UpdateErrorCodeEventHandler
    End Sub

    Protected Sub ReportMessage(message As String)
        OnStatusEvent(message)
    End Sub

    Protected Sub ReportError(message As String,
                              Optional eNewErrorCode As eMasicErrorCodes = eMasicErrorCodes.NoError)

        If eNewErrorCode <> eMasicErrorCodes.NoError Then
            OnUpdateErrorCode(eNewErrorCode, False)
        End If

        OnErrorEvent(message)
    End Sub

    Protected Sub ReportError(message As String,
                              ex As Exception,
                              Optional eNewErrorCode As eMasicErrorCodes = eMasicErrorCodes.NoError)

        If eNewErrorCode <> eMasicErrorCodes.NoError Then
            OnUpdateErrorCode(eNewErrorCode, False)
        End If

        OnErrorEvent(message, ex)
    End Sub

    <Obsolete("Source, allowInformUser, and allowThrowException are no longer supported")>
    Protected Sub ReportError(
      source As String,
      message As String,
      ex As Exception,
      allowInformUser As Boolean,
      Optional allowThrowException As Boolean = True,
      Optional eNewErrorCode As eMasicErrorCodes = eMasicErrorCodes.NoError)

        ReportError(message, ex, eNewErrorCode)
    End Sub

    Protected Sub ReportWarning(message As String)
        OnWarningEvent(message)
    End Sub

    Protected Sub SetBaseClassErrorCode(eNewErrorCode As ProcessFilesErrorCodes)
        OnUpdateBaseClassErrorCode(eNewErrorCode)
    End Sub

    Protected Sub SetLocalErrorCode(eNewErrorCode As eMasicErrorCodes, Optional leaveExistingErrorCodeUnchanged As Boolean = False)
        OnUpdateErrorCode(eNewErrorCode, leaveExistingErrorCodeUnchanged)
    End Sub

    Protected Sub UpdateCacheStats(objSpectraCache As clsSpectraCache)
        OnUpdateCacheStats(objSpectraCache.CacheEventCount, objSpectraCache.UnCacheEventCount)
    End Sub

    ''' <summary>
    ''' Update the progress of a given subtask
    ''' </summary>
    ''' <param name="percentComplete"></param>
    Protected Sub UpdateProgress(percentComplete As Short)
        OnProgressUpdate("", percentComplete)
    End Sub

    Protected Sub UpdateProgress(progressMessage As String)
        OnProgressUpdate(progressMessage, mLastPercentComplete)
    End Sub

    Protected Sub UpdateProgress(percentComplete As Short, progressMessage As String)
        mLastPercentComplete = percentComplete
        OnProgressUpdate(progressMessage, percentComplete)
    End Sub


#Region "Event Handlers"

    Private Sub UpdatedCacheStatsEventHandler(cacheEventCount As Integer, unCacheEventCount As Integer)
        OnUpdateCacheStats(cacheEventCount, unCacheEventCount)
    End Sub

    Private Sub UpdateBaseClassErrorCodeEventHandler(eErrorCode As ProcessFilesErrorCodes)
        SetBaseClassErrorCode(eErrorCode)
    End Sub

    Private Sub UpdateErrorCodeEventHandler(eErrorCode As eMasicErrorCodes, leaveExistingErrorCodeUnchanged As Boolean)
        SetLocalErrorCode(eErrorCode, leaveExistingErrorCodeUnchanged)
    End Sub
#End Region

End Class

File: clsMasicEventNotifier.cs
using System;
using PRISM;

namespace MASIC
{
    /// <summary>
    /// MASIC event notifier
    /// </summary>
    public abstract class clsMasicEventNotifier : EventNotifier
    {
        // Ignore Spelling: uncache

        private short mLastPercentComplete;

        /// <summary>
        /// Provides information on the number of cache and uncache events in spectraCache
        /// </summary>
        public event UpdateCacheStatsEventEventHandler UpdateCacheStatsEvent;

        /// <summary>
        /// Delegate for the update cache stats event
        /// </summary>
        /// <param name="cacheEventCount"></param>
        /// <param name="unCacheEventCount"></param>
        /// <param name="spectraPoolHitEventCount"></param>
        public delegate void UpdateCacheStatsEventEventHandler(int cacheEventCount, int unCacheEventCount, int spectraPoolHitEventCount);

        /// <summary>
        /// Update the code associated with an error
        /// </summary>
        public event UpdateBaseClassErrorCodeEventEventHandler UpdateBaseClassErrorCodeEvent;

        /// <summary>
        /// Delegate for the base class error code event
        /// </summary>
        /// <param name="eNewErrorCode"></param>
        public delegate void UpdateBaseClassErrorCodeEventEventHandler(PRISM.FileProcessor.ProcessFilesBase.ProcessFilesErrorCodes eNewErrorCode);

        /// <summary>
        /// Update the code associated with an error
        /// </summary>
        public event UpdateErrorCodeEventEventHandler UpdateErrorCodeEvent;

        /// <summary>
        /// Delete fro the update error code event
        /// </summary>
        /// <param name="eNewErrorCode"></param>
        /// <param name="leaveExistingErrorCodeUnchanged"></param>
        public delegate void UpdateErrorCodeEventEventHandler(clsMASIC.MasicErrorCodes eNewErrorCode, bool leaveExistingErrorCodeUnchanged);

        private void OnUpdateCacheStats(int cacheEventCount, int unCacheEventCount, int spectraPoolHitEventCount)
        {
            UpdateCacheStatsEvent?.Invoke(cacheEventCount, unCacheEventCount, spectraPoolHitEventCount);
        }

        private void OnUpdateBaseClassErrorCode(PRISM.FileProcessor.ProcessFilesBase.ProcessFilesErrorCodes eNewErrorCode)
        {
            UpdateBaseClassErrorCodeEvent?.Invoke(eNewErrorCode);
        }

        private void OnUpdateErrorCode(clsMASIC.MasicErrorCodes eNewErrorCode, bool leaveExistingErrorCodeUnchanged)
        {
            UpdateErrorCodeEvent?.Invoke(eNewErrorCode, leaveExistingErrorCodeUnchanged);
        }

        /// <summary>
        /// Register events
        /// </summary>
        /// <param name="sourceClass"></param>
        protected void RegisterEvents(clsMasicEventNotifier sourceClass)
        {
            base.RegisterEvents(sourceClass);

            sourceClass.UpdateCacheStatsEvent += UpdatedCacheStatsEventHandler;
            sourceClass.UpdateBaseClassErrorCodeEvent += UpdateBaseClassErrorCodeEventHandler;
            sourceClass.UpdateErrorCodeEvent += UpdateErrorCodeEventHandler;
        }

        /// <summary>
        /// Report a status message
        /// </summary>
        /// <param name="message"></param>
        protected void ReportMessage(string message)
        {
            OnStatusEvent(message);
        }

        /// <summary>
        /// Report an error message
        /// </summary>
        /// <param name="message"></param>
        /// <param name="eNewErrorCode"></param>
        protected void ReportError(string message,
                                   clsMASIC.MasicErrorCodes eNewErrorCode = clsMASIC.MasicErrorCodes.NoError)
        {
            if (eNewErrorCode != clsMASIC.MasicErrorCodes.NoError)
            {
                OnUpdateErrorCode(eNewErrorCode, false);
            }

            OnErrorEvent(message);
        }

        /// <summary>
        /// Report an error message, including an exception
        /// </summary>
        /// <param name="message"></param>
        /// <param name="ex"></param>
        /// <param name="eNewErrorCode"></param>
        protected void ReportError(string message,
                                   Exception ex,
                                   clsMASIC.MasicErrorCodes eNewErrorCode = clsMASIC.MasicErrorCodes.NoError)
        {
            if (eNewErrorCode != clsMASIC.MasicErrorCodes.NoError)
            {
                OnUpdateErrorCode(eNewErrorCode, false);
            }

            OnErrorEvent(message, ex);
        }

        /// <summary>
        /// Report a warning message
        /// </summary>
        /// <param name="message"></param>
        protected void ReportWarning(string message)
        {
            OnWarningEvent(message);
        }

        /// <summary>
        /// Set the base class error code
        /// </summary>
        /// <param name="eNewErrorCode"></param>
        protected void SetBaseClassErrorCode(PRISM.FileProcessor.ProcessFilesBase.ProcessFilesErrorCodes eNewErrorCode)
        {
            OnUpdateBaseClassErrorCode(eNewErrorCode);
        }

        /// <summary>
        /// Set a local error code
        /// </summary>
        /// <param name="eNewErrorCode"></param>
        /// <param name="leaveExistingErrorCodeUnchanged"></param>
        protected void SetLocalErrorCode(clsMASIC.MasicErrorCodes eNewErrorCode, bool leaveExistingErrorCodeUnchanged = false)
        {
            OnUpdateErrorCode(eNewErrorCode, leaveExistingErrorCodeUnchanged);
        }

        /// <summary>
        /// Update cache stats
        /// </summary>
        /// <param name="spectraCache"></param>
        protected void UpdateCacheStats(clsSpectraCache spectraCache)
        {
            OnUpdateCacheStats(spectraCache.CacheEventCount, spectraCache.UnCacheEventCount, spectraCache.SpectraPoolHitEventCount);
        }

        /// <summary>
        /// Update the progress of a given subtask
        /// </summary>
        /// <param name="percentComplete"></param>
        protected void UpdateProgress(short percentComplete)
        {
            OnProgressUpdate(string.Empty, percentComplete);
        }

        /// <summary>
        /// Update progress
        /// </summary>
        /// <param name="progressMessage"></param>
        protected void UpdateProgress(string progressMessage)
        {
            OnProgressUpdate(progressMessage, mLastPercentComplete);
        }

        /// <summary>
        /// Update progress
        /// </summary>
        /// <param name="percentComplete"></param>
        /// <param name="progressMessage"></param>
        protected void UpdateProgress(short percentComplete, string progressMessage)
        {
            mLastPercentComplete = percentComplete;
            OnProgressUpdate(progressMessage, percentComplete);
        }

        private void UpdatedCacheStatsEventHandler(int cacheEventCount, int unCacheEventCount, int spectraPoolHitEventCount)
        {
            OnUpdateCacheStats(cacheEventCount, unCacheEventCount, spectraPoolHitEventCount);
        }

        private void UpdateBaseClassErrorCodeEventHandler(PRISM.FileProcessor.ProcessFilesBase.ProcessFilesErrorCodes eErrorCode)
        {
            SetBaseClassErrorCode(eErrorCode);
        }

        private void UpdateErrorCodeEventHandler(clsMASIC.MasicErrorCodes eErrorCode, bool leaveExistingErrorCodeUnchanged)
        {
            SetLocalErrorCode(eErrorCode, leaveExistingErrorCodeUnchanged);
        }
    }
}
